File format security issues

derobins opened the pull request:
Addresses file format security issues detected via fuzzing.

Credit: Amazon Web Services and @qkoziol

Commit:
Patch H5encode.h

Commit:
Patched H5Zscaleoffset.c

Commit:
H5O patches (partial)

Commit:
Patch H5Pencdec.c and related

Commit:
Patch H5HG.c

Commit:
Patch H5Rint.c

Commit:
Format source

Commit:
Add H5F RFIC boilerplate

Commit:
Add H5P RFIC code

Commit:
Add H5T_is_numeric_with_unusual_unused_bits()

Commit:
Add H5_RANGE_OVERLAP()

Commit:
Patch H5S

Commit:
Patch H5Olinfo.c

Commit:
Patch H5A and related

Commit:
Patch H5Dio.c

Commit:
Format source

Commit:
Patch h5tools_str.c

Commit:
Patch H5Odtype.c

Commit:
Patch H5Tcommit.c

Commit:
Patch tmisc.c test 22

Commit:
Add unused bits checks and tools patches

NOTE: Requires fixup!
* Sizes are tweaked by one byte for unknown reasons
* tmisc_38 still barfs

Commit:
Patch H5Tconv.c

Commit:
Format source

Commit:
Update overflow macro

Commit:
Merge remote-tracking branch 'canonical/develop' into aws_13m_security_fixes

Commit:
Fix TRACE line

Commit:
Add tmisc38 binary files & enable test

Commit:
Fix spelling

derobins commented:
This needs RELEASE.txt info, which we'll add shortly. Let's get this review started so we can get this into develop and hdf5_1_14 ASAP.

mattjala commented on src/H5Spoint.c:
```diff
@@ -576,8 +576,14 @@ H5S__point_add(H5S_t *space, H5S_seloper_t op, size_t num_elem, const hsize_t *c
         /* Allocate space for the new node */
-        if (NULL == (new_node = (H5S_pnt_node_t *)H5FL_ARR_MALLOC(hcoords_t, space->extent.rank)))
+        /* Note: allocating "rank + 1" to allow for 'next' pointer */
+        if (NULL == (new_node = (H5S_pnt_node_t *)H5FL_ARR_MALLOC(hcoords_t, space->extent.rank + 1)))
```
How is this allocation not causing constant memory errors in the current version of the library due to being too small?

jhendersonHDF replied:
It's likely due to memory that was already allocated on the free list being reused here

derobins commented:
I have a set of files to add to https://github.com/HDFGroup/cve_hdf5, which I will do after we merge these changes to hdf5_1_14. If I do it now, the CI will break.

jhendersonHDF commented on src/H5Odtype.c:
```diff
@@ -452,6 +487,13 @@ H5O__dtype_decode_helper(unsigned *ioflags /*in,out*/, const uint8_t **pp, H5T_t
                 }
                 if (temp_type->shared->size == 0)
                     HGOTO_ERROR(H5E_DATATYPE, H5E_CANTDECODE, FAIL, "type size can't be zero");
+                if ((dt->shared->u.compnd.memb[dt->shared->u.compnd.nmembs].offset +
```
Seems like this should be `dt->shared->u.compnd.nmembs - 1`

jhendersonHDF replied:
Actually this one's fine. `dt->shared->u.compnd.nmembs` is the loop index here so this should be valid.

jhendersonHDF commented on src/H5Olinfo.c:
```diff
@@ -140,6 +140,9 @@ H5O__linfo_decode(H5F_t *f, H5O_t H5_ATTR_UNUSED *open_oh, unsigned H5_ATTR_UNUS
         INT64DECODE(p, linfo->max_corder);
+        if (linfo->max_corder < 0)
+            HGOTO_ERROR(H5E_OHDR, H5E_BADVALUE, NULL, "invalid max creation order value for message: %lld",
+                        (long long)linfo->max_corder);
```
Can be printed with `message: %" PRId64` here rather than casting

derobins replied:
Fixed

jhendersonHDF commented on test/tmisc.c:
Should probably use "tmisc38c.h5" for this filename just to avoid any test weirdness in the future

byrnHDF replied:
Oh, it will overwrite the B file and not create a C, right?

jhendersonHDF replied:
Looks like it from the test

derobins replied:
Fixed

jhendersonHDF commented on tools/test/h5dump/expected/tnbit.ddl:
```diff
@@ -1,10 +1,10 @@
    DATASPACE  SIMPLE { ( 20, 10 ) / ( 20, 10 ) }
    STORAGE_LAYOUT {
       CHUNKED ( 10, 5 )
-      SIZE XXXX (10.XXX:1 COMPRESSION)
```
Likely still want to leave the size and compression ratio masked out here

derobins replied:
Fixed

jhendersonHDF commented on src/H5Pfapl.c:
```diff
@@ -6264,3 +6276,72 @@ H5P__facc_vol_close(const char H5_ATTR_UNUSED *name, size_t H5_ATTR_UNUSED size,
+    H5TRACE2("e", "i*UL", plist_id, flags);
+
+    /* Get the property list structure */
+    if (NULL == (plist = H5P_object_verify(plist_id, H5P_FILE_ACCESS)))
```
Not a necessity, but it'd be nice to account for `H5P_DEFAULT` here

jhendersonHDF commented on src/H5Ppublic.h:
```diff
@@ -5634,6 +5634,79 @@ H5_DLL herr_t H5Pset_mdc_image_config(hid_t plist_id, H5AC_cache_image_config_t
+ *          metadata, allowing accidental data corruption to be correctly
+ *          detected and errors correctly issued without ambiguity.
+ *
+ * \since 1.14.3
```
1.14.4

derobins replied:
Fixed

jhendersonHDF commented on src/H5Ppublic.h:
```diff
@@ -5634,6 +5634,79 @@ H5_DLL herr_t H5Pset_mdc_image_config(hid_t plist_id, H5AC_cache_image_config_t
+ *          integrity check value into \p flags for the file access property
+ *          list specified in \p plist_id.
+ *
+ * \since 1.14.3
```
Also 1.14.4

derobins replied:
Fixed

jhendersonHDF commented on test/tmisc.c:
```diff
@@ -6264,14 +6287,230 @@ test_misc37(void)
+        if (u > 0) {
+            CHECK(ret, FAIL, "H5Tcommit2");
+
+            ret = H5Tclose(tid);
```
Should move this outside the if - else block so we close `tid` even if `H5Tcommit2` fails above

derobins replied:
Done

bmribler commented on src/H5Tcommit.c:
```diff
@@ -510,23 +526,39 @@ H5T__commit(H5F_t *file, H5T_t *type, hid_t tcpl_id)
+            if (H5O_close(oloc_ptr, NULL) < 0)
                 HDONE_ERROR(H5E_DATATYPE, H5E_CLOSEERROR, FAIL, "unable to release object header");
-            if (H5O_delete(file, type->sh_loc.u.loc.oh_addr) < 0)
+            if (H5O_delete(file, oloc_ptr->addr) < 0)
```
Can oloc_ptr->addr still be used after H5O_close(oloc_ptr, NULL)?

derobins replied:
No, it points to temp_oloc, which is local and cleaned up immediately after this.

jhendersonHDF commented on src/H5Tconv.c:
```diff
@@ -3801,6 +3818,57 @@ H5T__conv_vlen(const H5T_t *src, const H5T_t *dst, H5T_cdata_t *cdata, const H5T
+            while (dest_count > 0) {
+                H5_GCC_DIAG_OFF("discarded-qualifiers")
+                H5_CLANG_DIAG_OFF("incompatible-pointer-types-discards-qualifiers")
+                H5T_vlen_reclaim_elmt(d, dst); /* Ignore errors at this point */
```
Rather than hiding the const warnings here due to `H5T_vlen_reclaim_elmt` taking a `H5T_t *` parameter, I think we can just update `H5T_vlen_reclaim_elmt` to take a `const H5T_t *` instead. All it does is pass that parameter to `H5T__vlen_reclaim`, which takes a `const H5T_t *`. Would be much cleaner.

derobins replied:
Fixed

jhendersonHDF commented on src/H5Tconv.c:
```diff
@@ -3801,6 +3818,57 @@ H5T__conv_vlen(const H5T_t *src, const H5T_t *dst, H5T_cdata_t *cdata, const H5T
+                H5_GCC_DIAG_OFF("discarded-qualifiers")
+                H5_CLANG_DIAG_OFF("incompatible-pointer-types-discards-qualifiers")
+                H5T_vlen_reclaim_elmt(d, dst); /* Ignore errors at this point */
+                H5_GCC_DIAG_ON("discarded-qualifiers")
```
Same comment here about `H5T_vlen_reclaim_elmt`

derobins replied:
Fixed

jhendersonHDF commented on src/H5Tconv.c:
```diff
@@ -4221,6 +4303,49 @@ H5T__conv_ref(const H5T_t *src, const H5T_t *dst, H5T_cdata_t *cdata,
 
 done:
+    /* Release converted elements on error */
+    if (ret_value < 0 && conversions_made) {
```
After looking at this and the above side by side, it looks like the only difference is whether we call `H5T_vlen_reclaim_elmt` or `H5Rdestroy` on the element. Would be much nicer inside a helper function.

derobins replied:
Agree, but I'll leave this for later work. Right now I just want to get this in.

jhendersonHDF commented on src/H5Tconv.c:
```diff
@@ -3801,6 +3818,57 @@ H5T__conv_vlen(const H5T_t *src, const H5T_t *dst, H5T_cdata_t *cdata, const H5T
+            else
+                d = (uint8_t *)buf + (nelmts * orig_d_stride);
+
+            /* Destroy references that have already been converted */
```
Nitpick: this is for freeing variable-length elements rather than destroying references. Same for below

derobins replied:
Fixed

jhendersonHDF commented on src/H5private.h:
```diff
@@ -338,7 +347,10 @@
+#define H5_IS_BUFFER_OVERFLOW(ptr, size, buffer_end)                                                         \
+    (((ptr) > (buffer_end)) ||                             /* Bad precondition */                            \
+     ((ptrdiff_t)(size) > (((buffer_end) - (ptr)) + 1)) || /* Typical overflow */                            \
+     ((intptr_t)(size) < 0))                               /* Negative 'size' would wrap 'ptr' */
```
This is the only part of the PR I keep getting hung up on. If a large enough value is passed in for `size`, won't this be implementation-defined what happens if it overflows when casting to `intptr_t`? It seems like we need something like:

`((size <= (size_t)INTPTR_MAX) && ((intptr_t)(size) < 0))`

though, I'm not quite sure that's right either in the case where `INTPTR_MAX` is larger than `SIZE_MAX`, causing wraparound when casting `INTPTR_MAX` to `size_t`

fortnern replied:
So I think it's technically possible to have a valid buffer so large that ptrdiff_t wraps to negative when computing the size. This could be an issue on 32 bit systems that are pushing the limits of their architecture. So size could be negative when cast to ptrdiff_t and still be valid. I don't know if the rest of the library would handle this case correctly.

fortnern replied:
Maybe something like `(((ptr) > (buffer_end)) || (((ptr) + (size) - 1) > (buffer_end)) || (((ptr) + (size)) < (ptr)))`

We might want to cast the pointers to `const uint8_t *` as well, and size to `size_t`.

jhendersonHDF replied:
Hmm yes I think this is a bit more in line with what I was thinking.

jhendersonHDF replied:
After a little research though, it seems some compilers often remove a `ptr + len < ptr` type check since it's UB anyway.

qkoziol replied:
I suggest getting my notes for files that tripped overflows from Dana and making certain that any change to what I worked out last year are still protecting from the related corrupted files.  Some of them were pretty obscure... :-/

jhendersonHDF replied:
Definitely plan to test against the CVE files. I think it's mostly a matter of needing to check `size` against the max of the signed type it's being cast into first, and then casting it and checking against < 0 before then doing the pointer subtraction and comparison. Unfortunately it's difficult to do this in a nice way since the macro has no type information and it's not always a `size_t` value that's being passed in.

jhendersonHDF replied:
Rewrote this slightly. It's basically the same thing, but using `ptrdiff_t` instead of `intptr_t`, checking against `PTRDIFF_MAX` before casting and also doing the check for negative values before the typical overflow check. Passes all current CVE tests, including the ones specifically around the changes in this PR.

Because of the "type-limits" warnings that had to be suppressed in this PR, I'm tempted to just move the macro into an inline function, but for the sake of getting these changes in, I don't think needs to happen right now.

jhendersonHDF commented on src/H5T.c:
```diff
@@ -6638,3 +6639,43 @@ H5T__get_path_table_npaths(void)
+ *-------------------------------------------------------------------------
+ */
+bool
+H5T_is_numeric_with_unusual_unused_bits(const H5T_t *dt)
```
Testing on PPC64LE by @hyoklee shows that this probably won't work correctly for the IBM long double type. Suggest either:

```C
#if LDBL_MANT_DIG == 106
    /* This won't work for the IBM long double type */
    HGOTO_DONE(false);
#else
    ...
#endif
```

or some other solution, until we can maybe address issues with the datatype.

Commit:
Fix format specifier

Commit:
New functions since 1.14.4 in Doxygen

Commit:
Mask out nbit h5dump test

Commit:
vlen const changes in H5T

Commit:
Reference --> vlen in a comment

Commit:
ID cleanup in tmisc38

Commit:
Fix tmisc38 filename C

Commit:
Merge remote-tracking branch 'canonical/develop' into aws_13m_security_fixes

jhendersonHDF committed:
Ensure ref. priv. buffer pointers are aligned

jhendersonHDF committed:
Handle H5P_DEFAULT in H5Pget_relax_file_integrity_checks

jhendersonHDF committed:
Handle IBM long double type in H5T_is_numeric_with_unusual_unused_bits

jhendersonHDF committed:
Suppress type-limits warnings for Clang as well

jhendersonHDF commented on src/H5encode.h:
```diff
@@ -224,7 +224,7 @@
         (p) += 8;                                                                                            \
         for (_i = 0; _i < sizeof(uint64_t); _i++)                                                            \
-            n = (n << 8) | *(--p);                                                                           \
+            n = (uint64_t)(((uint64_t)n << 8) | *(--p));                                                     \
```
Casting this back to a `uint64_t` seems to be showing a warning:

```
conversion from ‘long unsigned int’ to ‘uint32_t’ {aka ‘unsigned int’} may change value [-Warith-conversion]
  227 |             n = (uint64_t)(((uint64_t)n << 8) | *(--p));                                                     \
      |                 ^
src/H5Olayout.c:398:33: note: in expansion of macro ‘H5F_DECODE_LENGTH’
  398 |                                 H5F_DECODE_LENGTH(f, p, mesg->storage.u.chunk.u.single.nbytes);
      |                                 ^~~~~~~~~~~~~~~~~
```

Seems like we could end up trying to UINT64DECODE into a `uint32_t` value

jhendersonHDF replied:
Updated this to decode into a uint64_t and then checked_assign it to the uint32_t field.

jhendersonHDF committed:
Address warning in H5Olayout.c

jhendersonHDF committed:
Rewrite H5_IS_BUFFER_OVERFLOW slightly

check value against PTRDIFF_MAX before casting to ptrdiff_t

update H5_IS_KNOWN_BUFFER_OVERFLOW macro to use H5_IS_BUFFER_OVERFLOW

github-actions[bot] committed:
Committing clang-format changes

jhendersonHDF committed:
Initialize var.

jhendersonHDF commented:
CMake seems to be failing in the `H5DUMP-tnbit` test because the compression ratio line isn't getting masked out for some reason. Not sure on netcdf failures, but investigating.

jhendersonHDF commented:
Updating the `H5_IS_KNOWN_BUFFER_OVERFLOW` macro to use `H5_IS_BUFFER_OVERFLOW` when `skip` is false caused the netCDF tests to start failing in one test case when trying to open a file after creating a dataset with an opaque type in it. This fails both before and after my changes to `H5_IS_BUFFER_OVERFLOW`. More investigation is likely needed there, but I'll put the change back for now.

jhendersonHDF committed:
Revert change to H5_IS_KNOWN_BUFFER_OVERFLOW

fortnern commented on src/H5Odtype.c:
```diff
@@ -176,6 +176,14 @@ H5O__dtype_decode_helper(unsigned *ioflags /*in,out*/, const uint8_t **pp, H5T_t
+            /* Sanity checks */
+            if (dt->shared->u.atomic.offset >= (dt->shared->size * 8))
+                HGOTO_ERROR(H5E_DATATYPE, H5E_BADRANGE, FAIL, "integer offset out of bounds");
+            if (0 == dt->shared->u.atomic.prec)
```
Could we support this as a no-op?

fortnern commented on src/H5Odtype.c:
```diff
@@ -176,6 +176,14 @@ H5O__dtype_decode_helper(unsigned *ioflags /*in,out*/, const uint8_t **pp, H5T_t
+                HGOTO_ERROR(H5E_DATATYPE, H5E_BADRANGE, FAIL, "integer offset out of bounds");
+            if (0 == dt->shared->u.atomic.prec)
+                HGOTO_ERROR(H5E_DATATYPE, H5E_BADVALUE, FAIL, "precision is zero");
+            if (((dt->shared->u.atomic.offset + dt->shared->u.atomic.prec) - 1) >= (dt->shared->size * 8))
```
Could we get rid of the -1 and change the >= to a > here (and in other places in this function)?

fortnern commented on src/H5Odtype.c:
```diff
@@ -452,6 +487,13 @@ H5O__dtype_decode_helper(unsigned *ioflags /*in,out*/, const uint8_t **pp, H5T_t
                 }
                 if (temp_type->shared->size == 0)
                     HGOTO_ERROR(H5E_DATATYPE, H5E_CANTDECODE, FAIL, "type size can't be zero");
+                if ((dt->shared->u.compnd.memb[dt->shared->u.compnd.nmembs].offset +
```
I think we should either always use `offset + size > dt->shared->size` or `offset + size - 1 >= dt->shared->size`

fortnern commented on src/H5Oattribute.c:
```diff
@@ -211,6 +211,13 @@ H5O__attr_create(const H5O_loc_t *loc, H5A_t *attr)
         HGOTO_ERROR(H5E_ATTR, H5E_CANTPIN, FAIL, "unable to pin object header");
 
+    /* Check for creating attribute with unusual datatype */
+    if (H5T_is_numeric_with_unusual_unused_bits(attr->shared->dt) &&
```
Maybe move the checksum part up so we can short circuit the more expensive function call

fortnern commented on src/H5Odtype.c:
```diff
@@ -770,6 +812,19 @@ H5O__dtype_decode_helper(unsigned *ioflags /*in,out*/, const uint8_t **pp, H5T_t
     }
 
+    /* Check for numeric type w/unusual # of unused bits */
+    if (H5T_is_numeric_with_unusual_unused_bits(dt))
```
This doesn't seem to suppress the error when checksumming is turned on

jhendersonHDF committed:
Mask out compression filter in CMake for tnbit test

derobins commented:
> CMake seems to be failing in the `H5DUMP-tnbit` test because the compression ratio line isn't getting masked out for some reason. Not sure on netcdf failures, but investigating.

In the Autotools, I had to change the lead digit of a macro, since it goes from 10 to 1.

jhendersonHDF commented:
Will address the H5_IS_BUFFER_OVERFLOW issue and @fortnern's comments in a separate PR so as to not hold up this PR and create a bunch of extra testing work.
